File: dataset_reduzido/avatar/Python/Code/atcoder_ABC143_D.py
from bisect import bisect_right , bisect_left
n , * l = map ( int , open ( 0 ).read ( ).split ( ) )
l.sort ( )
ans = 0
for i in range ( n ) :
    for j in range ( i + 1 , n ) :
        a = l [ i ]
        b = l [ j ]
        right = bisect_left ( l , a + b )
        left = bisect_right ( l , max ( a - b , b - a ) )
        tmp = max ( 0 , right - left )
        if left <= i < right :
            tmp -= 1
        if left <= j < right :
            tmp -= 1
        ans += tmp
print ( ans // 3 )

File: output/Qwen_repaired_1/avatar/Python/Java/atcoder_ABC143_D.java
import java.util.*;
import java.io.*;

public class atcoder_ABC143_D {
    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        String[] input = reader.readLine().split(" ");
        int n = Integer.parseInt(input[0]);
        List<Integer> list = new ArrayList<>();
        for (int i = 1; i < input.length; ++i) {
            list.add(Integer.parseInt(input[i]));
        }
        Collections.sort(list);
        long ans = 0;
        for (int i = 0; i < n; ++i) {
            for (int j = i + 1; j < n; ++j) {
                int a = list.get(i);
                int b = list.get(j);
                int left = bisectRight(list, Math.max(a - b, b - a));
                int right = bisectLeft(list, a + b);
                if (left <= i && i < right) --right;
                if (left <= j && j < right) --right;
                ans += Math.max(0, right - left);
            }
        }
        System.out.println(ans / 3);
    }

    private static int bisectRight(List<Integer> list, int key) {
        int low = 0, high = list.size();
        while (low < high) {
            int mid = (low + high) >>> 1;
            if (key < list.get(mid)) high = mid;
            else low = mid + 1;
        }
        return low;
    }

    private static int bisectLeft(List<Integer> list, int key) {
        int low = 0, high = list.size();
        while (low < high) {
            int mid = (low + high) >>> 1;
            if (key <= list.get(mid)) high = mid;
            else low = mid + 1;
        }
        return low;
    }
}